File: java/src/main/java/it/giannibombelli/wsc2026/booking/domain/events/BookingPlaced.java
package it.giannibombelli.wsc2026.booking.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;


public record BookingPlaced(BookingId aggregateId, Money amount, Description description,
                            GiftCardId giftCardId) implements BookingEvent {
    public BookingPlaced {
        Require.requireArgument(aggregateId, "bookingId");
        Require.requireArgument(amount, "amount");
        Require.requireArgument(description, "description");
        Require.requireArgument(giftCardId, "giftCardId");
    }
}

File: typescript/src/booking/domain/events/bookingPlaced.ts
// Evento emesso quando una prenotazione viene inserita.


import type { Event } from '@/common/domain/model/event.js';
import type { Description } from '@/common/domain/primitive/description.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { BookingId } from '../booking/bookingId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type BookingPlaced = Event<'BookingPlaced', BookingId> & {
  readonly amount: Money;
  readonly description: Description;
  readonly giftCardReference: string;
};

export function bookingPlaced(
  aggregateId: BookingId,
  amount: Money,
  description: Description,
  giftCardReference: string
): BookingPlaced {
  requireArgument(aggregateId, 'bookingId');
  requireArgument(amount, 'amount');
  requireArgument(description, 'description');
  requireArgument(giftCardReference, 'giftCardReference');
  return { kind: 'BookingPlaced', aggregateId, amount, description, giftCardReference };
}
